### PR TITLE
python: allow versions with garbage suffix

### PR DESCRIPTION
**Pull request by trws**
Ubuntu 22.04 preview python prints version as 3.10.2+, the + causes
version parsing to fail and breaks detection.

### CODE REVIEW

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -272,7 +272,7 @@ def determine_version(cls, exe):
         # Output looks like:
         #   Python 3.7.7
         output = Executable(exe)('-V', output=str, error=str)
-        match = re.search(r'Python\s+(\S+)', output)
+        match = re.search(r'Python\s+([0-9a-zA-Z.\-]+)', output)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py]: No need to escape the `-` if it comes last. Also, you can add `_` if you want to match `spack.versions.VALID_VERSION`.

File: var/spack/repos/builtin/packages/python/package.py [hunk 2]
```diff
@@ -272,7 +272,7 @@ def determine_version(cls, exe):
         # Output looks like:
         #   Python 3.7.7
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py:273]: Can you add a comment here with the output on your system?

Comment (trws) [reply var/spack/repos/builtin/packages/python/package.py:273]: Pre-patch this is the spack output, will add python in a sec, have to start the image back up:
```
root@ef9d30ceec49:/curdir# spack external find python
==> Warning: error detecting "python" from prefix /bin [Bad characters in version string: 3.10.2+]
==> Warning: error detecting "python" from prefix /usr/bin [Bad characters in version string: 3.10.2+]
==> No new external packages detected
```

Comment (trws) [reply var/spack/repos/builtin/packages/python/package.py:273]: Here's the python output:
```
root@0593d449d500:/curdir# python3 -V
Python 3.10.2+
```

File: var/spack/repos/builtin/packages/python/package.py (hunk 1 above)
Comment (trws) [reply var/spack/repos/builtin/packages/python/package.py]: Actually, I should probably just use VALID_VERSION, should have thought of that.

Comment (trws) [reply var/spack/repos/builtin/packages/python/package.py]: Ah, it clamps to the begin and end, that's inconvenient.

Comment (trws) [reply var/spack/repos/builtin/packages/python/package.py]: Reworked.

**Review by adamjstewart — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**